### PR TITLE
fix(install): frozen install no longer fails on cold cache with git apm_package deps

### LINKED ISSUE #2456: [BUG] apm install --frozen does not hydrate git apm_package dependencies on a cold cache, failing with "package manifest not found" when MCP state is present
**Describe the bug**

Since v0.27.0, `apm install --frozen` fails on a **cold cache** (a fresh checkout with an
empty/absent `apm_modules/`) for a git-hosted dependency whose `package_type` is
`apm_package` — i.e. a package that *does* ship an `apm.yml` — whenever the project also
carries MCP state. It exits with:

    --frozen: apm.lock.yaml is out of sync with apm.yml.
      - <host>/<org>/<repo>: package manifest not found at
        .../apm_modules/<host>/<org>/<repo>/apm.yml; re-run 'apm install' to restore it

The lockfile is not actually stale. A plain `apm install` on the same tree fetches the
dependency and then `apm install --frozen` passes; `apm audit` on the same tree reports no
drift. The failure is that `--frozen` never materialises (clones/checks out) the git
dependency on a cold cache before enforcing the check, so the manifest it looks for was
simply never fetched.

This is the standard CI shape: a fresh clone runs `apm install --frozen` directly, with no
prior plain install to populate `apm_modules/`.

**To Reproduce**

1. Create a consumer project whose `apm.yml` declares **one MCP server** and a git-hosted
   `apm_package` dependency (declared directly, or pulled in transitively by another git
   dependency — both reproduce). Example:

   ```yaml
   name: example-consumer
   version: 1.0.0
   targets:
     - claude
     - copilot

   dependencies:
     apm:
       - example.com/acme/example-skill#example-skill-v1.0.0
     mcp:
       - name: example-mcp
         registry: false
         transport: http
         url: https://mcp.example.com/mcp
   ```
   where `acme/example-skill` is an `apm_package` (its repo root has an `apm.yml`,
   `includes: auto`), tagged `example-skill-v1.0.0`.

2. Generate the lockfile once so it pins the dependency:

   ```
   apm install
   ```
   → Succeeds. The lockfile records `package_type: apm_package` for `example-skill`.

3. Simulate a fresh CI checkout — keep the committed `apm.yml` + `apm.lock.yaml`, remove the
   installed tree:

   ```
   rm -rf apm_modules
   apm install --frozen
   ```
   → Fails with `package manifest not found at .../apm_modules/example.com/acme/example-skill/apm.yml; re-run 'apm install' to restore it`.

4. Control — restore via a plain install, then re-run frozen, changing nothing else:

   ```
   apm install            # hydrates apm_modules/
   apm install --frozen   # now passes
   ```

**Expected behavior**

- On a cold cache, `apm install --frozen` should hydrate git `apm_package` dependencies from
  their locked pins (as `apm audit --ci` already self-hydrates), then validate — rather than
  reporting an absent, never-fetched manifest as lockfile drift.
- Equivalently: an absent `apm_modules/<dep>/apm.yml` for a locked `apm_package` git
  dependency on a cold cache should not be treated as "out of sync with apm.yml" when the
  lockfile pin is intact and a plain `apm install` would restore it.
- The advice `re-run 'apm install' to restore it` should not appear for a state where the
  frozen install itself is expected to be the restoring step (CI runs *only* `--frozen`).

**Environment (please complete the following information):**
 - OS: macOS 26.3.1 (arm64); also reproduces on Linux CI runners
 - Python Version: Homebrew-distributed prebuilt binary (PyInstaller bundle)
 - APM Version: 0.27.0 (703dd9e)
 - VSCode Version (if relevant): N/A

**Logs**

Cold-cache `apm install --frozen` (fresh checkout, empty `apm_modules/`):

    [>] Installing dependencies from apm.yml...
    [x] --frozen: apm.lock.yaml is out of sync with apm.yml.
      - example.com/acme/example-skill: package manifest not found at
        .../apm_modules/example.com/acme/example-skill/apm.yml; re-run 'apm install' to restore it
    [i] Tip: run 'apm outdated' to see what changed, then 'apm update'.
    [x] Install failed after 0.1s.

Same tree after a plain `apm install`, then `apm install --frozen`: passes, no changes.

**Additional context**

- This appears to share a regression origin with #2443: PR #2390 (v0.27.0) wired the MCP
  config view into `_enforce_frozen`, and that view walks the `apm.yml` of every locked
  dependency. The check only runs when MCP state is present
  (`bool(root_mcp) or lockfile.mcp_configs or lockfile.mcp_servers`), so a project with no
  MCP server does not hit it.
- It is **not** a duplicate of #2443: that issue covers manifestless repo-root `claude_skill`
  packages (which never ship an `apm.yml`, directory present). This case is an `apm_package`
  dependency (which *does* ship an `apm.yml`) whose directory is **absent** on a cold cache.
  The cold-cache fallback added in #2329 only exempts `claude_skill`, so `apm_package` git
  deps are still reported as missing.
- It is also distinct from #1996 (a fixed SSH↔normalized-URL matching false positive).
- Impact: breaks the CI-safe `apm install --frozen` path on every fresh checkout for any
  project that combines an MCP server with a git-hosted `apm_package` dependency. The only
  in-project workaround is to run a plain `apm install` (or restore the manifests) before the
  frozen gate, which defeats the purpose of `--frozen` as the sole CI install step.

### PR DESCRIPTION
**Pull request by sergio-sisternes-epam**
# fix(install): frozen install no longer fails on cold cache with git apm_package deps

apm-spec-waiver: regression fix restoring pre-v0.27.0 frozen-install semantics; no new observable behaviour added to the install or integration paths

## TL;DR

`apm install --frozen` failed with spurious lockfile-drift errors on a fresh checkout when the
project declared git `apm_package` dependencies. The MCP config check wired in v0.27.0 (#2390)
walked each locked dep's `apm.yml` before those packages were fetched — no file, no parse,
immediate failure. Two targeted fixes restore pre-0.27.0 semantics without relaxing any real
drift detection.

> [!NOTE]
> Closes #2456. No behaviour change for warm-cache installs or local `apm_package` deps.

## Problem (WHY)

- [x] `apm install --frozen` on a fresh checkout (`apm_modules/` empty) emits `McpSourceProblem`
  for every git `apm_package` dep and exits non-zero — hydration never runs.
- [x] `_allows_missing_manifest` returns `False` for `apm_package` types before checking whether
  the directory exists, because the `is_virtual_subdirectory()` guard fires first. The
  absent-directory branch at the bottom of the function is unreachable for this type.
- [!] When an absent git `apm_package` dep declares its own MCP servers, those servers appear in
  `lockfile.mcp_configs` but not in `current_mcp.configs`, producing false `lock_only` drift
  entries that compound the failure.

The regression was introduced in [PR #2390](https://github.com/microsoft/apm/pull/2390), which
wired `CurrentMcpConfigView.derive` into `enforce_frozen` for MCP config drift detection.
Cold-cache hydration is a valid, expected use of `--frozen` (CI first-run, clean dev machine
setup) and must not fail before packages are fetched.

## Approach (WHAT)

| # | Fix |
|---|-----|
| 1 | In `_allows_missing_manifest`, add an early-return **before** `is_virtual_subdirectory()`: absent + non-local + `APM_PACKAGE` → `True`. Mirrors the existing `claude_skill` cold-cache exemption in the same function. |
| 2 | New helper `_absent_apm_package_mcp_configs()` in `service.py`: uses `mcp_config_provenance` to find servers whose every owner is an absent non-local dep, then augments `current_mcp.configs` before computing the diff — both sides carry the same values for cold-cache targets, so no false `lock_only` entries. |

## Implementation (HOW)

- **`src/apm_cli/integration/mcp_config_view.py`** — 13-line early-return in
  `_allows_missing_manifest` placed before `is_virtual_subdirectory()`. Local `apm_package`
  deps (path-anchored, must always exist) are deliberately not exempted.
- **`src/apm_cli/install/service.py`** — New module-level `_absent_apm_package_mcp_configs(lockfile, modules_dir)`
  helper; `enforce_frozen` now builds `effective_configs` and calls
  `McpConfigDiff.between(effective_configs, lockfile.mcp_configs)` instead of `current_mcp.diff()`.
  The `expected_names` guard is updated in parallel.
- **`tests/unit/install/test_frozen.py`** — `TestEnforceFrozenColdCache`: 3 regression tests
  covering the primary scenario, the MCP-augmentation edge case, and a guard that real drift from
  an installed dep still fails.
- **`tests/unit/integration/test_mcp_config_view.py`** — 2 new tests: absent non-local
  `apm_package` dep emits no `McpSourceProblem`; absent local dep still records a problem.

## Diagrams

Execution flow through `enforce_frozen` after the fix. Yellow blocks are the two new code paths.

```mermaid
sequenceDiagram
    participant EF as enforce_frozen
    participant CMV as CurrentMcpConfigView.derive
    participant AMM as _allows_missing_manifest
    participant AH as _absent_apm_package_mcp_configs

    EF->>CMV: derive(request, lockfile, modules_dir)
    CMV->>AMM: check each locked dep
    rect rgb(255, 247, 200)
        Note over AMM: NEW: absent non-local apm_package returns True
        AMM-->>CMV: True (cold-cache exemption)
    end
    CMV-->>EF: current_mcp (no McpSourceProblem)
    rect rgb(255, 247, 200)
        EF->>AH: lockfile, modules_dir
        AH-->>EF: cold_cache_mcp (servers from absent deps)
        EF->>EF: effective_configs = current_mcp.configs merged cold_cache_mcp
        EF->>EF: McpConfigDiff.between(effective_configs, lockfile.mcp_configs)
    end
    EF-->>EF: no false lock_only entries
```

## Trade-offs

- **Augmentation uses `mcp_config_provenance` as the ownership signal, not source paths.** This
  is what the lock writer recorded at install time. If provenance is absent (pre-0.27.0 lockfile),
  `_absent_apm_package_mcp_configs` returns `{}` — no augmentation, no regression for old lockfiles.
- **Shared servers (co-owned by present and absent deps) are not augmented.** Only servers whose
  every owner is absent are included; co-owned servers remain subject to live comparison.
- **Local `apm_package` deps are not exempted.** Path-anchored; an absent directory is a
  genuine misconfiguration, not cold cache.
- **No schema, lockfile, or manifest format changes.** The fix is entirely in the read path of
  `enforce_frozen`.

## Benefits

1. `apm install --frozen` on a fresh CI checkout with git `apm_package` deps succeeds and
   hydrates correctly instead of failing before any work is done.
2. Real MCP config drift in an already-installed dep still fails frozen as expected — the guard
   is narrowed, not removed.
3. The `_allows_missing_manifest` exemption follows the same pattern already in the function
   (the existing `claude_skill` cold-cache exemption), keeping the logic consistent.
4. No CHANGELOG entry added for this fix; semantics are a pure regression restore and the entry
   will be cut with the next release.

## Validation

All lint checks pass:

```
ruff check src/ tests/    -- All checks passed!
ruff format --check       -- 1609 files already formatted
pylint R0801              -- Your code has been rated at 10.00/10
auth-signal lint          -- [+] auth-signal lint clean
```

<details><summary>pytest output — 32 tests, 0 failures</summary>

```
uv run --extra dev pytest tests/unit/install/test_frozen.py \
  tests/unit/integration/test_mcp_config_view.py -q

................................
32 passed in 0.43s
```

</details>

### Scenario Evidence

| # | Scenario (user promise) | Principle(s) | Test(s) proving it | Type |
|---|------------------------|--------------|--------------------|------|
| 1 | `apm install --frozen` on a fresh checkout with a git `apm_package` dep passes without error | Governed by policy, DevX | `tests/unit/install/test_frozen.py::TestEnforceFrozenColdCache::test_cold_cache_git_apm_package_dep_passes` (regression-trap for #2456) | unit |
| 2 | `apm install --frozen` where the absent git dep declares its own MCP servers passes (servers restored from lockfile) | Governed by policy, DevX | `tests/unit/install/test_frozen.py::TestEnforceFrozenColdCache::test_cold_cache_git_apm_package_with_own_mcp_passes` | unit |
| 3 | Real MCP drift in an already-installed dep still causes `--frozen` to fail | Governed by policy | `tests/unit/install/test_frozen.py::TestEnforceFrozenColdCache::test_cold_cache_does_not_mask_real_mcp_drift_from_installed_dep` | unit |
| 4 | Absent non-local `apm_package` dep emits no `McpSourceProblem` in `CurrentMcpConfigView.derive` | Governed by policy | `tests/unit/integration/test_mcp_config_view.py::test_absent_git_apm_package_dep_is_skipped_on_cold_cache` | unit |
| 5 | Absent local `apm_package` dep (path-anchored) still records a problem | Governed by policy, Secure by default | `tests/unit/integration/test_mcp_config_view.py::test_absent_local_apm_package_dep_still_records_problem` | unit |

## How to test

- [ ] Clone the repo fresh (`apm_modules/` empty) on a project with a git `apm_package` dep and
  MCP state in `apm.lock.yaml`; run `apm install --frozen` — expect success and hydration.
- [ ] On a warm cache, run `apm install --frozen` and confirm it still passes with no regressions.
- [ ] Edit an installed dep's `apm.yml` MCP config (real drift) and confirm `apm install --frozen`
  still exits non-zero.
- [ ] `uv run --extra dev pytest tests/unit/install/test_frozen.py tests/unit/integration/test_mcp_config_view.py -v` — all 32 pass.

Co-authored-by: Copilot <223556219+Copilot@users.noreply.github.com>

### CODE REVIEW

File: src/apm_cli/install/service.py [hunk 1]
```diff
@@ -240,18 +293,26 @@ def enforce_frozen(request: InstallRequest) -> None:
...
             )
-            config_diff = current_mcp.diff(lockfile.mcp_configs)
+            # On a cold cache, absent non-local apm_package deps have not been
+            # fetched yet and contribute nothing to current_mcp.configs, even
+            # though their MCP servers are stored in lockfile.mcp_configs. Augment
+            # the derived current state with those stored configs so the comparison
+            # produces no false lock_only entries for cold-cache hydration targets.
+            cold_cache_mcp = _absent_apm_package_mcp_configs(lockfile, modules_dir)
+            effective_configs: dict = {**current_mcp.configs, **cold_cache_mcp}
+            config_diff = McpConfigDiff.between(effective_configs, lockfile.mcp_configs)
```
[the remark this answers is not in the document]
Comment (sergio-sisternes-epam) [reply src/apm_cli/install/service.py:326]: Good catch. Flipped the merge order to `{**cold_cache_mcp, **current_mcp.configs}` in commit `189901881` so a live (present) server config always takes precedence over the lockfile-stored fallback when both sides carry the same server name.


> [!NOTE]
> Auto-resolved by the [GitHub Copilot app](https://github.com/github/app).

File: src/apm_cli/install/service.py [hunk 2]
```diff
@@ -22,16 +22,69 @@
...
+    from apm_cli.models.validation import PackageType
+
+    absent_declarers: set[str] = set()
+    for dep in lockfile.dependencies.values():
+        if dep.package_type != PackageType.APM_PACKAGE.value:
+            continue
+        if dep.source == "local":
+            continue
+        install_path = dep.to_dependency_ref().get_install_path(modules_dir)
+        if not install_path.exists():
```
[the remark this answers is not in the document]
Comment (sergio-sisternes-epam) [reply src/apm_cli/install/service.py]: Applied in commit `189901881`. Added `modules_dir_exists = modules_dir.exists()` once at the top of the function; the per-dep loop now checks `not modules_dir_exists or not install_path.exists()` — when the whole `apm_modules/` directory is absent, every remote dep is immediately classified as absent without individual stat calls.


> [!NOTE]
> Auto-resolved by the [GitHub Copilot app](https://github.com/github/app).

**Review by danielmeppiel — APPROVED** [review]

**Comment by danielmeppiel** [comment]
## APM Review Panel Advisory -- PR #2502

**Iteration 1 of 1 | Panel execution: inline**

---

### CEO Arbitration

Ship with follow-ups: community regression fix restores frozen-install on cold cache, well-tested and well-scoped.

This PR from a community contributor (sergio-sisternes-epam) restores pre-v0.27.0 behavior for `apm install --frozen` when the local cache is cold and dependencies include git-sourced apm_packages with MCP server declarations. The fix is surgically scoped, architecturally sound, and backed by five unit tests with mutation-break validation. All six active panelists converged: ship it, with only nit-tier and recommended-tier findings -- no blocking defects, no security concerns, no performance regressions.

DX-3 (verbose breadcrumb on rehydration) references `commands/install_cmd.py`, which is NOT modified by this PR -- scoped out. TC-1 (integration fixture tier test) belongs in a follow-up issue. PA-5 (co-owned server test) was folded in this run. GH-1 (CHANGELOG) was folded.

### Principle Alignment

The fix restores a broken user promise (frozen install with cold cache) without expanding scope, aligning with APM's install-contract integrity and the principle that regressions are communicated and resolved fast.

---

### Persona Summary

| Persona | Summary |
|---------|---------|
| python-architect | Clean regression fix; dict merge and provenance correlation are correct. One missing co-owned-server test (PA-5) is the only real gap -- folded. |
| cli-logging-expert | No user-facing output regressions; two minor formatting/observability nits on the new frozen-install error path. |
| devx-ux-expert | Frozen-install cold-cache fix is solid DX; the silent rehydration matches npm ci mental model. Minor polish on messaging. |
| supply-chain-security-expert | No new trust boundaries, no token exposure, no path traversal. Lockfile remains the integrity root. Clean from supply-chain perspective. |
| oss-growth-hacker | Solid CI-friction fix from a community contributor; CHANGELOG entry folded. |
| test-coverage-expert | Bug fix has 5 passing unit tests with mutation-break guard; co-owned server regression trap added. |
| performance-expert | No perf concern. New code is O(n+m) in-memory + at most n stats, called once per frozen install. Sub-5ms cost. |
| auth-expert | Inactive -- PR does not touch auth.py, token_manager.py, or auth-adjacent files. |

---

### Folded in this run

- **GH-1** (CHANGELOG entry): Added `[Unreleased] Fixed` entry for frozen-install cold-cache regression -- resolved in `01077c7`
- **PA-5** (co-owned MCP server test): Added `test_co_owned_server_not_exempted_by_cold_cache` regression-trapping the `issubset` guard in `_absent_apm_package_mcp_configs`; mutation-break validated -- resolved in `01077c7`
- **CI-TIMEOUT** (root-cause merge-queue fix): Raised integration shard timeout from 15 to 20 min; shard 1 holds all `test_architecture_authorities.py` tests (~29 copytree+bash guard tests, 50-60s each); 15min budget had no headroom causing intermittent merge-queue failures (observed: 15min21s on slow runner vs 13min35s on fast runner) -- resolved in `4b1b9c3`

### Copilot Signals Reviewed

- **Copilot inline 1** (dict merge collision): LEGIT -- already resolved in PR commit `e5d827e` (refactor: flip merge order to `{**cold_cache_mcp, **current_mcp.configs}` so live config wins)
- **Copilot inline 2** (modules_dir early exit): LEGIT -- already resolved in PR commit `e5d827e` (refactor: hoist `modules_dir_exists = modules_dir.exists()` to skip per-dep stat on cold cache)

### Deferred

- **TC-1** (integration-tier fixture test): Recommend filing a follow-up issue. Unit coverage is solid; integration-with-fixtures tier is a larger ask.
- **DX-3** (verbose breadcrumb on rehydration): Out of scope -- references `commands/install_cmd.py` which this PR does not change.
- **CL-2** (verbose trace for cold-cache skip): Enhancement, not a bug. Defer to follow-up.

### Lint Evidence

```
ruff check src/ tests/    -- All checks passed!
ruff format --check       -- 1610 files already formatted
pylint R0801              -- Your code has been rated at 10.00/10
auth-signal lint          -- [+] auth-signal lint clean
boundary lint             -- [+] architecture boundary lint clean
```

### CI Evidence

All 16 PR-head checks GREEN on `01077c7`:
Lint pass 2m49s | Build & Test Shard 1 pass 2m27s | Build & Test Shard 2 pass 2m39s |
Test Architecture Ratchets pass 1m55s | Lifecycle Smoke pass 3m2s | gate pass 3m11s |
Windows Compatibility Gate pass 1m27s | CodeQL pass | Spec conformance pass | all others pass.

### Mergeability

| #PR | sha | CEO stance | iter | folds | deferrals | copilot_rounds | ci_status | mergeable | merge_state_status | notes |
|-----|-----|-----------|------|-------|-----------|----------------|-----------|-----------|-------------------|-------|
| #2502 | 01077c7 | ship_with_followups | 1 | 3 | 3 | 1 | green | MERGEABLE | BLOCKED | awaiting maintainer re-approval |

---

*Shepherd-driver advisory -- do NOT auto-merge. Parent will re-approve and enqueue.*

**Comment by rrazvd** [comment]
@sergio-sisternes-epam hello, do we have dependencies to merge this PR?

**Comment by sergio-sisternes-epam** [comment]
> @sergio-sisternes-epam hello, do we have dependencies to merge this PR?

No blocking dependencies. But I cannot approve my own PR. I need @danielmeppiel to get ths approved.
